File: TypeScript/src/dfa.ts
import { DFAValidationError } from "./errors";
import type { DFAConfig, DFAResult, ExecutionTrace } from "./types";

export class DFA {
	private readonly states: Set<string>;
	private readonly alphabet: Set<string>;
	private readonly transitions: Map<string, Map<string, string>>;
	private readonly startState: string;
	private readonly acceptingStates: Set<string>;

	constructor(config: DFAConfig) {
		this.states = new Set(config.states);
		this.alphabet = new Set(config.alphabet);
		this.transitions = this.buildTransitionMap(config.transitions);
		this.startState = config.startState;
		this.acceptingStates = new Set(config.acceptingStates);

		this.validate();
	}

	private buildTransitionMap(
		transitions: Record<string, Record<string, string>>,
	): Map<string, Map<string, string>> {
		const map = new Map<string, Map<string, string>>();

		for (const [state, stateTransitions] of Object.entries(transitions)) {
			map.set(state, new Map(Object.entries(stateTransitions)));
		}

		return map;
	}

	private validate(): void {
		// check start state belongs to the Q
		if (!this.states.has(this.startState)) {
			throw new DFAValidationError(
				`Start state "${this.startState}" is not in the set of states`,
			);
		}

		// check accepting / final states is a subset to the Q
		for (const acceptingState of this.acceptingStates) {
			if (!this.states.has(acceptingState)) {
				throw new DFAValidationError(
					`Accepting state "${acceptingState}" is not in the set of states`,
				);
			}
		}

		// check transitions are total: every state has transitions for every symbol
		for (const state of this.states) {
			const stateTransitions = this.transitions.get(state);
			if (stateTransitions === undefined) {
				throw new DFAValidationError(
					`State "${state}" has no transitions defined`,
				);
			}

			for (const symbol of this.alphabet) {
				const targetState = stateTransitions.get(symbol);
				if (targetState === undefined) {
					throw new DFAValidationError(
						`Missing transition for state "${state}" on symbol "${symbol}"`,
					);
				}

				if (!this.states.has(targetState)) {
					throw new DFAValidationError(
						`Transition from "${state}" on "${symbol}" leads to unknown state "${targetState}"`,
					);
				}
			}
		}
	}

	run(input: string): DFAResult {
		let currentState = this.startState;

		for (const symbol of input) {
			if (!this.alphabet.has(symbol)) {
				return "rejected";
			}

			const stateTransitions = this.transitions.get(currentState);
			const nextState = stateTransitions?.get(symbol);
			if (nextState === undefined) {
				return "rejected";
			}

			currentState = nextState;
		}

		return this.acceptingStates.has(currentState) ? "accepted" : "rejected";
	}

	runWithTrace(input: string): ExecutionTrace {
		const steps: ExecutionTrace["steps"] = [];
		let currentState = this.startState;

		for (const symbol of input) {
			const stateTransitions = this.transitions.get(currentState);
			const nextState = stateTransitions?.get(symbol);
			if (!this.alphabet.has(symbol) || nextState === undefined) {
				return {
					input,
					startState: this.startState,
					steps,
					finalState: currentState,
					result: "rejected",
				};
			}

			steps.push({
				fromState: currentState,
				toState: nextState,
				symbol,
			});

			currentState = nextState;
		}

		return {
			input,
			startState: this.startState,
			steps,
			finalState: currentState,
			result: this.acceptingStates.has(currentState) ? "accepted" : "rejected",
		};
	}
}

File: Rust/src/dfa.rs
//! DFA implementation with validation.

use std::collections::{HashMap, HashSet};

use crate::error::ValidationError;
use crate::types::{Config, ExecutionStep, ExecutionTrace};

#[derive(Debug, Clone)]
pub struct DFA {
    states: HashSet<String>,
    alphabet: HashSet<char>,
    transitions: HashMap<String, HashMap<char, String>>,
    start_state: String,
    accepting_states: HashSet<String>,
}

impl DFA {
    pub fn new(config: Config) -> Result<Self, ValidationError> {
        let dfa = Self {
            states: config.states.into_iter().collect(),
            alphabet: config.alphabet.into_iter().collect(),
            transitions: config.transitions,
            start_state: config.start_state,
            accepting_states: config.accepting_states.into_iter().collect(),
        };

        dfa.validate()?;
        Ok(dfa)
    }

    fn validate(&self) -> Result<(), ValidationError> {
        if !self.states.contains(&self.start_state) {
            return Err(ValidationError::InvalidStartState(self.start_state.clone()))
        }

        for state in &self.accepting_states {
            if !self.states.contains(state) {
                return Err(ValidationError::InvalidAcceptingState(state.clone()));
            }
        }

        for state in &self.states {
            let Some(state_transitions) = self.transitions.get(state) else {
                return Err(ValidationError::MissingStateTransitions(state.clone()));
            };

            for symbol in &self.alphabet {
                let Some(target) = state_transitions.get(symbol) else {
                    return Err(ValidationError::MissingTransition {
                        state: state.clone(),
                        symbol: *symbol,
                    });
                };

                if !self.states.contains(target) {
                    return Err(ValidationError::InvalidTransitionTarget {
                        from: state.clone(),
                        symbol: *symbol,
                        to: target.clone(),
                    });
                }
            }
        }

        Ok(())
    }

    pub fn run(&self, input: &str) -> bool {
        let mut current_state = &self.start_state;

        for symbol in input.chars() {
            if !self.alphabet.contains(&symbol) {
                return false;
            }

            let Some(state_transitions) = self.transitions.get(current_state) else {
                return false;
            };

            let Some(next_state) = state_transitions.get(&symbol) else {
                return false;
            };

            current_state = next_state;
        }

        self.accepting_states.contains(current_state)
    }

    pub fn run_with_trace(&self, input: &str) -> ExecutionTrace {
        let mut steps = Vec::new();
        let mut current_state = &self.start_state;

        for symbol in input.chars() {
            if !self.alphabet.contains(&symbol) {
                return ExecutionTrace {
                    input: input.to_string(),
                    start_state: self.start_state.clone(),
                    steps,
                    final_state: current_state.clone(),
                    result: false,
                };
            }

            let Some(state_transitions) = self.transitions.get(current_state) else {
                return ExecutionTrace {
                    input: input.to_string(),
                    start_state: self.start_state.clone(),
                    steps,
                    final_state: current_state.clone(),
                    result: false,
                };
            };

            let Some(next_state) = state_transitions.get(&symbol) else {
                return ExecutionTrace {
                    input: input.to_string(),
                    start_state: self.start_state.clone(),
                    steps,
                    final_state: current_state.clone(),
                    result: false,
                };
            };

            steps.push(ExecutionStep { from_state: current_state.clone(), to_state: next_state.clone(), symbol, });

            current_state = next_state;
        }

        ExecutionTrace {
            input: input.to_string(),
            start_state: self.start_state.clone(),
            steps,
            final_state: current_state.clone(),
            result: self.accepting_states.contains(current_state),
        }
    }
}

